README
Apache Evasive Maneuvers Module
For Apache 1.3, 2.0 and 2.4
Jonathan Zdziarski
Version 1.10 [2005.0117]

LICENSE

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; version 2
of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

WHAT IS MOD_EVASIVE ?

mod_evasive is an evasive maneuvers module for Apache to provide evasive
action in the event of an HTTP DoS or DDoS attack or brute force attack.  It 
is also designed to be a detection tool, and can be easily configured to talk 
to ipchains, firewalls, routers, and etcetera.  

Detection is performed by creating an internal dynamic hash table of IP 
Addresses and URIs, and denying any single IP address from any of the following:

- Requesting the same page more than a few times per second
- Making more than 50 concurrent requests on the same child per second
- Making any requests while temporarily blacklisted (on a blocking list)

This method has worked well in both single-server script attacks as well 
as distributed attacks, but just like other evasive tools, is only as 
useful to the point of bandwidth and processor consumption (e.g. the
amount of bandwidth and processor required to receive/process/respond
to invalid requests), which is why it's a good idea to integrate this
with your firewalls and routers.

This module instantiates for each listener individually, and therefore has
a built-in cleanup mechanism and scaling capabilities.  Because of this,
legitimate requests are rarely ever compromised, only legitimate attacks.  Even
a user repeatedly clicking on 'reload' should not be affected unless they do
it maliciously.

Four different module sources have been provided:

Apache v1.3 API:	mod_evasive14.c
Apache v2.0 API:	mod_evasive20.c
Apache v2.4 API:	mod_evasive24.c
NSAPI (iPlanet):	mod_evasiveNSAPI.c *

NOTE: mod_evasiveNSAPI is a port submitted by Reine Persson <reiper@rsv.se>
      and is not officially supported as part of the mod_evasive project.

HOW IT WORKS

A web hit request comes in. The following steps take place:

- The IP address of the requestor is looked up on the temporary blacklist
- The IP address of the requestor and the URI are both hashed into a "key".  
  A lookup is performed in the listener's internal hash table to determine 
  if the same host has requested this page more than once within the past 
  1 second.  
- The IP address of the requestor is hashed into a "key".
  A lookup is performed in the listerner's internal hash table to determine
  if the same host has requested more than 50 objects within the past
  second (from the same child).

If any of the above are true, a 403 response is sent.  This conserves
bandwidth and system resources in the event of a DoS attack.  Additionally,
a system command and/or an email notification can also be triggered to block
all the originating addresses of a DDoS attack. 

Once a single 403 incident occurs, mod_evasive now blocks the entire IP 
address for a period of 10 seconds (configurable).  If the host requests a 
page within this period, it is forced to wait even longer.  Since this is 
triggered from requesting the same URL multiple times per second, this 
again does not affect legitimate users.

The blacklist can/should be configured to talk to your network's firewalls 
and/or routers to push the attack out to the front lines, but this is not 
required.

mod_evasive also performs syslog reporting using daemon.alert.  Messages
will look like this:

Aug  6 17:41:49 elijah mod_evasive[23184]: [ID 801097 daemon.alert] Blacklisting address x.x.x.x: possible attack.

WHAT IS THIS TOOL USEFUL FOR?

This tool is *excellent* at fending off request-based DoS attacks or scripted
attacks, and brute force attacks. When integrated with firewalls or IP filters,
mod_evasive can stand up to even large attacks. Its features will prevent you 
from wasting bandwidth or having a few thousand CGI scripts running as a 
result of an attack.  

If you do not have an infrastructure capable of fending off any other types
of DoS attacks, chances are this tool will only help you to the point of
your total bandwidth or server capacity for sending 403's.  Without a solid
infrastructure and address filtering tool in place, a heavy distributed DoS 
will most likely still take you offline.  

HOW TO INSTALL

APACHE v1.3
-----------

Without DSO Support:

1. Extract this archive into src/modules in the Apache source tree

2. mv mod_evasive14.c mod_evasive.c

3. Run ./configure --add-module=src/modules/evasive/mod_evasive.c

4. make, install

5. Restart Apache 

With DSO Support, Ensim, or CPanel:

1. $APACHE_ROOT/bin/apxs -iac mod_evasive.c

2. Restart Apache

APACHE v2.0/v2.4
----------------

1. Extract this archive

2. Rename the version you need (mv mod_evasive24.c mod_evasive.c)

3. Run $APACHE_ROOT/bin/apxs -i -a -c mod_evasive.c

4. The module will be built and installed into $APACHE_ROOT/modules, and loaded into your httpd.conf

5. Restart Apache

NSAPI
SunONE (iPlanet,netscape) Installation
--------------------------------------

Tested on:
iPlanet 4.1sp12
iPlanet 6.0sp5

Edit compile script for your environment and compile mod_evasiveNSAPI.c
to a shared library.

CONFIGURATION

mod_evasive has default options configured, but you may also add the
following block to your httpd.conf:

APACHE
------

<IfModule mod_evasive.c>
    DOSHashTableSize    3097
    DOSPageCount        2
    DOSSiteCount        50
    DOSPageInterval     1
    DOSSiteInterval     1
    DOSBlockingPeriod   10
</IfModule>

Optionally you can also add the following directives:

    DOSEmailNotify	you@yourdomain.com
    DOSSystemCommand	"su - someuser -c '/sbin/... %s ...'"
    DOSLogDir		"/var/lock/mod_evasive"
		DOSHTTPStatus 429

You will also need to add this line if you are building with dynamic support:

APACHE v1.3
-----------

AddModule	mod_evasive.c

APACHE v2.0
-----------

LoadModule evasive_module modules/mod_evasive.so

(This line is already added to your configuration by apxs)

NSAPI
SunONE (iPlanet,Netscape) Configuration
--------------------------------------
                                                                                
Configure iPlanet 4.1
---------------------

Edit obj.conf:
                                                                                
Init fn="load-modules" funcs="mod_evasive_init,mod_evasive_check" shlib="/opt/ns-4.1/plugins/lib/mod_evasive.sl"
                                                                                
Init fn="mod_evasive_init" DOSPageCount=2 DOSSiteCount=50 DOSPageInterval=1 DOSSiteInterval=1 DOSBlockingPeriod=10 DOSWhitelist="10.60.0.7,10.65.0.10"
                                                                                
In the default object:
PathCheck fn=mod_evasive_check
                                                                                
Or an own object
<Object name="evasive" ppath="/DoSProtectedArea*">
NameTrans fn=mod_evasive_check
</Object>
                                                                                
                                                                                
Configure iPlanet 6.0
---------------------
                                                                                
Edit magnus.conf:
                                                                                
Init fn="load-modules" funcs="mod_evasive_init,mod_evasive_check" shlib="/opt/iplanet-6.0/plugins/lib/mod_evasive.sl"
                                                                                
Init fn="mod_evasive_init" DOSWhitelist="10.60.0.7,10.65.0.10"
                                                                                
Edit obj.conf:
In the default object:
PathCheck fn=mod_evasive_check
                                                                                
Or an own object
<Object name="evasive" ppath="/DoSProtectedArea*">
NameTrans fn=mod_evasive_check
</Object>

DOSHashTableSize
----------------

The hash table size defines the number of top-level nodes for each child's 
hash table.  Increasing this number will provide faster performance by 
decreasing the number of iterations required to get to the record, but 
consume more memory for table space.  You should increase this if you have
a busy web server.  The value you specify will automatically be tiered up to 
the next prime number in the primes list (see mod_evasive.c for a list 
of primes used).

DOSPageCount
------------

This is the threshhold for the number of requests for the same page (or URI)
per page interval.  Once the threshhold for that interval has been exceeded,
the IP address of the client will be added to the blocking list.
 
DOSSiteCount
------------

This is the threshhold for the total number of requests for any object by
the same client on the same listener per site interval.  Once the threshhold 
for that interval has been exceeded, the IP address of the client will be added
to the blocking list.

DOSPageInterval
---------------

The interval for the page count threshhold; defaults to 1 second intervals.

DOSSiteInterval
---------------

The interval for the site count threshhold; defaults to 1 second intervals.

DOSBlockingPeriod
-----------------

The blocking period is the amount of time (in seconds) that a client will be
blocked for if they are added to the blocking list.  During this time, all
subsequent requests from the client will result in a 403 (Forbidden) and
the timer being reset (e.g. another 10 seconds).  Since the timer is reset
for every subsequent request, it is not necessary to have a long blocking
period; in the event of a DoS attack, this timer will keep getting reset. 

DOSEmailNotify
--------------

If this value is set, an email will be sent to the address specified
whenever an IP address becomes blacklisted.  A locking mechanism using /tmp
prevents continuous emails from being sent.

NOTE: Be sure MAILER is set correctly in mod_evasive.c. The default is
			"/bin/mail -t %s" where %s is used to denote the destination email
			address set in the configuration. If you are running on linux or some
			other operating system with a different type of mailer, you'll need to
			change this.

DOSSystemCommand
----------------

If this value is set, the system command specified will be executed
whenever an IP address becomes blacklisted.  This is designed to enable
system calls to ip filter or other tools.  A locking mechanism using /tmp
prevents continuous system calls.  Use %s to denote the IP address of the
blacklisted IP.

DOSLogDir
---------

Choose an alternative temp directory

By default "/tmp" will be used for locking mechanism, which opens some 
security issues if your system is open to shell users.

  	http://security.lss.hr/index.php?page=details&ID=LSS-2005-01-01

In the event you have nonprivileged shell users, you'll want to create a
directory writable only to the user Apache is running as (usually root),
then set this in your httpd.conf.

DOSHTTPStatus
-------------

Choose an alternative HTTP status code for the reply to blocked clients.

By default mod_evasive returns 403 Forbidden to blocked clients. This
directive allows any other HTTP code known to Apache to be used instead.

WHITELISTING IP ADDRESSES

IP addresses of trusted clients can be whitelisted to insure they are never 
denied.  The purpose of whitelisting is to protect software, scripts, local 
searchbots, or other automated tools from being denied for requesting large 
amounts of data from the server.  Whitelisting should *not* be used to add 
customer lists or anything of the sort, as this will open the server to abuse.
This module is very difficult to trigger without performing some type of 
malicious attack, and for that reason it is more appropriate to allow the 
module to decide on its own whether or not an individual customer should be 
blocked.

To whitelist an address (or range) add an entry to the Apache configuration 
in the following fashion:

DOSWhitelist	127.0.0.1
DOSWhitelist	127.0.0.*

Wildcards can be used on up to the last 3 octets if necessary.  Multiple
DOSWhitelist commands may be used in the configuration.

TWEAKING APACHE

The keep-alive settings for your children should be reasonable enough to 
keep each child up long enough to resist a DOS attack (or at least part of 
one).  Remember, it is the child processes that maintain their own internal
IP address tables, and so when one exits, so does all of the IP information it
had. For every child that exits, another 5-10 copies of the page may get 
through before putting the attacker back into '403 Land'.  With this said, 
you should have a very high MaxRequestsPerChild, but not unlimited as this
will prevent cleanup.

You'll want to have a MaxRequestsPerChild set to a non-zero value, as
DosEvasive cleans up its internal hashes only on exit.  The default
MaxRequestsPerChild is usually 10000.  This should suffice in only allowing
a few requests per 10000 per child through in the event of an attack (although
if you use DOSSystemCommand to firewall the IP address, a hole will no
longer be open in between child cycles).

TESTING

Want to make sure it's working? Run test.pl, and view the response codes.
It's best to run it several times on the same machine as the web server until
you get 403 Forbidden messages. Some larger servers with high child counts 
may require more of a beating than smaller servers before blacklisting
addresses. 

Please don't use this script to DoS others without their permission.

KNOWN BUGS

- This module appears to conflict with the Microsoft Frontpage Extensions.
  Frontpage sucks anyway, so if you're using Frontpage I assume you're asking
  for problems, and not really interested in conserving server resources anyway.

FEEDBACK 

Please email me with questions, constructive comments, or feedback:
  jonathan@nuclearelephant.com



CHANGELOG
Version 2.2.0
-----------

[20200208] jvdmr: Add DOSWhitelistUri config option

Version 2.1.1
-----------

[20171011] jvdmr: Set Debian package version

Version 2.0
-----------

[20171010] jvdmr: Added Windows support (based off Greg Smith's port)

[20171010] jvdmr: Renamed mod_evasive14 to mode_evasive13 to match Apache 1.3 version number

[20171010] jvdmr: Backported new features to Apache 2.0

[20171010] jvdmr: Debian package for Apache 2.4

[20171010] jvdmr: Support for per-vhost configuration instead of only global

[20171009] jvdmr: evasive20_module is now simply called evasive_module

[20171009] jvdmr: Added DOSHTTPStatus directive

[20171009] jvdmr: Added support for Apache 2.4

Version 1.10.1
--------------

[20051008] jonz: Fixed IP Whitelisting in Apache 1.3 Version

[20051008] jonz: Corrected initialization to prevent dumping IP database

[20051008] jonz: Documentation and code cleaned up, renamed mod_evasive

Version 1.10.0
--------------

[20050117] jonz: Security fix: Tempdir configuration directive (race condition)

Version 1.9.0
-------------

[20031030] jonz: Added NSAPI/SunONE Support

[20031030] jonz: Added TEMP_HOME definition to change temporary file locations

Version 1.8.0
-------------

[20030901] jonz: Added support for IP Whitelisting

Version 1.7.1
-------------

[20030826] jonz: Minor bugfixes to Apache 2.0 module

[20030826] jonz: Corrections to object creation and cleanup

Version 1.7.0
-------------

[20030822] jonz: Support for Apache 2.0

Version 1.6.1
-------------

[20030811] jonz: Bugfix: free() of another static variable
 
Version 1.6.0
-------------

[20030806] jonz: Added syslog support

Version 1.5.1
-------------

[20030516] jonz: Bugfix: free() of a static variable

Version 1.5.0
-------------

[20030425] jonz: Added automated email notification 

[20030425] jonz: Added configurable system command

Version 1.4.0
-------------

[20021031] jonz: Added support fror httpd.conf directives

[20021031] jonz: Added --add-module support

Version 1.0.0
-------------

[20021015] jonz: Initial Release


mod_evasive.c
/* $Id: mod_evasive.c,v 1.3 2005/10/08 19:17:14 jonz Exp $ */

/*
mod_evasive for Apache 1.3
Copyright (c) by Jonathan A. Zdziarski

LICENSE

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; version 2
of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
                                                                                
*/

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <time.h>
#include <syslog.h>
#include <errno.h>

#include "httpd.h"
#include "http_core.h"
#include "http_config.h"
#include "http_log.h"
#include "http_request.h"

module MODULE_VAR_EXPORT evasive_module;

/* BEGIN DoS Evasive Maneuvers Definitions */

#define MAILER	"/bin/mail -t %s"
#define  LOG( A, ... ) { openlog("mod_evasive", LOG_PID, LOG_DAEMON); syslog( A, __VA_ARGS__ ); closelog(); }

#define DEFAULT_HASH_TBL_SIZE   3079ul  // Default hash table size
#define DEFAULT_PAGE_COUNT      2       // Default max page hit count/interval
#define DEFAULT_SITE_COUNT      50      // Default max site hit count/interval
#define DEFAULT_PAGE_INTERVAL   1       // Default 1 second page interval
#define DEFAULT_SITE_INTERVAL   1       // Default 1 second site interval
#define DEFAULT_BLOCKING_PERIOD 10      // Default block time (Seconds)
#define DEFAULT_LOG_DIR		"/tmp"

/* END DoS Evasive Maneuvers Definitions */

/* BEGIN NTT (Named Timestamp Tree) Headers */

enum { ntt_num_primes = 28 };

/* ntt root tree */
struct ntt {
    long size;
    long items;
    struct ntt_node **tbl;
};

/* ntt node (entry in the ntt root tree) */
struct ntt_node {
    char *key;
    time_t timestamp;
    long count;
    struct ntt_node *next;
};

/* ntt cursor */
struct ntt_c {
  long iter_index;
  struct ntt_node *iter_next;
};

struct ntt *ntt_create(long size);
struct ntt_node	*ntt_find(struct ntt *ntt, const char *key);
struct ntt_node	*ntt_insert(struct ntt *ntt, const char *key, time_t timestamp);
struct ntt_node *c_ntt_first(struct ntt *ntt, struct ntt_c *c);
struct ntt_node *c_ntt_next(struct ntt *ntt, struct ntt_c *c);
int ntt_destroy(struct ntt *ntt);
int ntt_delete(struct ntt *ntt, const char *key);
long ntt_hashcode(struct ntt *ntt, const char *key);

/* END NTT (Named Timestamp Tree) Headers */


/* BEGIN DoS Evasive Maneuvers Globals */

struct ntt *hit_list;	// Our dynamic hash table
struct ntt *white_list = NULL; // White list table

static unsigned long hash_table_size = DEFAULT_HASH_TBL_SIZE;
static int page_count      = DEFAULT_PAGE_COUNT;
static int page_interval   = DEFAULT_PAGE_INTERVAL;
static int site_count      = DEFAULT_SITE_COUNT;
static int site_interval   = DEFAULT_SITE_INTERVAL;
static int blocking_period = DEFAULT_BLOCKING_PERIOD;
static char *log_dir       = NULL;
static char *email_notify  = NULL;
static char *sys_command   = NULL;
int is_whitelisted(const char *ip);
static const char *whitelist(cmd_parms *cmd, void *mconfig, char *ip);

/* END DoS Evasive Maneuvers Globals */

static void evasive_child_init(server_rec *s, pool *p)
{
    hit_list   = ntt_create(hash_table_size);
}

static int check_access(request_rec *r) 
{
    int ret = OK;

    /* BEGIN Evasive Maneuvers Code */

    if (r->prev == NULL && r->main == NULL && hit_list != NULL) {
      unsigned long address = r->connection->remote_addr.sin_addr.s_addr;
      char *text_add = inet_ntoa(r->connection->remote_addr.sin_addr);
      char hash_key[2048];
      struct ntt_node *n;
      time_t t = time(NULL);

      /* Check whitelist */
       
      if (is_whitelisted(text_add))
        return OK;

      /* First see if the IP itself is on "hold" */
      snprintf(hash_key, 2048, "%ld", address);
      n = ntt_find(hit_list, hash_key);

      if (n != NULL && t-n->timestamp<blocking_period) {
 
        /* If the IP is on "hold", make it wait longer in 403 land */
        ret = FORBIDDEN;
        n->timestamp = time(NULL);

      /* Not on hold, check hit stats */
      } else {

        /* Has URI been hit too much? */
        snprintf(hash_key, 2048, "%ld_%s", address, r->uri);
        n = ntt_find(hit_list, hash_key);
        if (n != NULL) {

          /* If URI is being hit too much, add to "hold" list and 403 */
          if (t-n->timestamp<page_interval && n->count>=page_count) {
            ret = FORBIDDEN;
            snprintf(hash_key, 2048, "%ld", address);
            ntt_insert(hit_list, hash_key, time(NULL));
          } else {

            /* Reset our hit count list as necessary */
            if (t-n->timestamp>=page_interval) {
              n->count=0;
            }
          }
          n->timestamp = t;
          n->count++;
        } else {
          ntt_insert(hit_list, hash_key, t);
        }

        /* Has site been hit too much? */
        snprintf(hash_key, 2048, "%ld_SITE", address);
        n = ntt_find(hit_list, hash_key);
        if (n != NULL) {

          /* If site is being hit too much, add to "hold" list and 403 */
          if (t-n->timestamp<site_interval && n->count>=site_count) {
            ret = FORBIDDEN;
            snprintf(hash_key, 2048, "%ld", address);
            ntt_insert(hit_list, hash_key, time(NULL));
          } else {

            /* Reset our hit count list as necessary */
            if (t-n->timestamp>=site_interval) {
              n->count=0;
            }
          }
          n->timestamp = t;
          n->count++;
        } else {
          ntt_insert(hit_list, hash_key, t);
        }
      }

      /* Perform email notification and system functions */
      if (ret == FORBIDDEN) {
        char filename[1024];
        struct stat s;
        FILE *file;

        snprintf(filename, sizeof(filename), "%s/dos-%s", log_dir != NULL ? log_dir : DEFAULT_LOG_DIR, text_add);
        if (stat(filename, &s)) {
          file = fopen(filename, "w");
          if (file != NULL) {
            fprintf(file, "%ld\n", getpid());
            fclose(file);

            LOG(LOG_ALERT, "Blacklisting address %s: possible attack.", text_add)
            if (email_notify != NULL) {
              snprintf(filename, sizeof(filename), MAILER, email_notify);
              file = popen(filename, "w");
              if (file != NULL) {
                fprintf(file, "To: %s\n", email_notify);
                fprintf(file, "Subject: HTTP BLACKLIST %s\n\n", text_add);
                fprintf(file, "mod_evasive HTTP Blacklisted %s\n", text_add);
                pclose(file);
              }
            }

            if (sys_command != NULL) {
              snprintf(filename, sizeof(filename), sys_command, text_add);
              system(filename);
            }
 
          } else {
		LOG(LOG_ALERT, "Couldn't open logfile %s: %s",filename, strerror(errno));
	  }

        } /* if (temp file does not exist) */

      } /* if (ret == FORBIDDEN) */

    } /* if (r->prev == NULL && r->main == NULL && hit_list != NULL) */

    /* END Evasive Maneuvers Code */

    if (ret == FORBIDDEN
	&& (ap_satisfies(r) != SATISFY_ANY || !ap_some_auth_required(r))) {
	ap_log_rerror(APLOG_MARK, APLOG_NOERRNO|APLOG_ERR, r,
		  "client denied by server configuration: %s",
		  r->filename);
    }

    return ret;
}

static void evasive_child_exit(server_rec *s, pool *p) 
{
    ntt_destroy(hit_list);
    free(email_notify);
    free(sys_command);
}


/* BEGIN NTT (Named Timestamp Tree) Functions */

static unsigned long ntt_prime_list[ntt_num_primes] = 
{
    53ul,         97ul,         193ul,       389ul,       769ul,
    1543ul,       3079ul,       6151ul,      12289ul,     24593ul,
    49157ul,      98317ul,      196613ul,    393241ul,    786433ul,
    1572869ul,    3145739ul,    6291469ul,   12582917ul,  25165843ul,
    50331653ul,   100663319ul,  201326611ul, 402653189ul, 805306457ul,
    1610612741ul, 3221225473ul, 4294967291ul
};


/* Find the numeric position in the hash table based on key and modulus */

long ntt_hashcode(struct ntt *ntt, const char *key) {
    unsigned long val = 0;
    for (; *key; ++key) val = 5 * val + *key;
    return(val % ntt->size);
}

/* Creates a single node in the tree */

struct ntt_node *ntt_node_create(const char *key) {
    char *node_key;
    struct ntt_node* node;

    node = (struct ntt_node *) malloc(sizeof(struct ntt_node));
    if (node == NULL) {
	return NULL;
    }
    if ((node_key = strdup(key)) == NULL) {
        free(node);
	return NULL;
    }
    node->key = node_key;
    node->timestamp = time(NULL);
    node->next = NULL;
    return(node);
}

/* Tree initializer */

struct ntt *ntt_create(long size) {
    long i = 0;
    struct ntt *ntt = (struct ntt *) malloc(sizeof(struct ntt));

    if (ntt == NULL)
        return NULL;
    while (ntt_prime_list[i] < size) { i++; }
    ntt->size  = ntt_prime_list[i];
    ntt->items = 0;
    ntt->tbl   = (struct ntt_node **) calloc(ntt->size, sizeof(struct ntt_node *));
    if (ntt->tbl == NULL) {
        free(ntt);
        return NULL;
    }
    return(ntt);
}

/* Find an object in the tree */

struct ntt_node *ntt_find(struct ntt *ntt, const char *key) {
    long hash_code;
    struct ntt_node *node;

    if (ntt == NULL) return NULL;

    hash_code = ntt_hashcode(ntt, key);
    node = ntt->tbl[hash_code];

    while (node) {
        if (!strcmp(key, node->key)) {
            return(node);
        }
        node = node->next;
    }
    return((struct ntt_node *)NULL);
}

/* Insert a node into the tree */

struct ntt_node *ntt_insert(struct ntt *ntt, const char *key, time_t timestamp) {
    long hash_code;
    struct ntt_node *parent;
    struct ntt_node *node;
    struct ntt_node *new_node = NULL;

    if (ntt == NULL) return NULL;

    hash_code = ntt_hashcode(ntt, key);
    parent	= NULL;
    node	= ntt->tbl[hash_code];

    while (node != NULL) {
        if (strcmp(key, node->key) == 0) { 
            new_node = node;
            node = NULL;
        }

	if (new_node == NULL) {
          parent = node;
          node = node->next;
        }
    }

    if (new_node != NULL) {
        new_node->timestamp = timestamp;
        new_node->count = 0;
        return new_node; 
    }

    /* Create a new node */
    new_node = ntt_node_create(key);
    new_node->timestamp = timestamp;
    new_node->timestamp = 0;

    ntt->items++;

    /* Insert */
    if (parent) {  /* Existing parent */
	parent->next = new_node;
        return new_node;  /* Return the locked node */
    }

    /* No existing parent; add directly to hash table */
    ntt->tbl[hash_code] = new_node;
    return new_node;
}

/* Tree destructor */

int ntt_destroy(struct ntt *ntt) {
    struct ntt_node *node, *next;
    struct ntt_c c;

    if (ntt == NULL) return -1;

    node = c_ntt_first(ntt, &c);
    while(node != NULL) {
        next = c_ntt_next(ntt, &c);
        ntt_delete(ntt, node->key);
        node = next;
    }

    free(ntt->tbl);
    free(ntt);
    ntt = (struct ntt *) NULL;

    return 0;
}

/* Delete a single node in the tree */

int ntt_delete(struct ntt *ntt, const char *key) {
    long hash_code;
    struct ntt_node *parent = NULL;
    struct ntt_node *node;
    struct ntt_node *del_node = NULL;

    if (ntt == NULL) return -1;

    hash_code = ntt_hashcode(ntt, key);
    node        = ntt->tbl[hash_code];

    while (node != NULL) {
        if (strcmp(key, node->key) == 0) {
            del_node = node;
            node = NULL;
        }

        if (del_node == NULL) {
          parent = node;
          node = node->next;
        }
    }

    if (del_node != NULL) {

        if (parent) {
            parent->next = del_node->next;
        } else {
            ntt->tbl[hash_code] = del_node->next;
        }

        free(del_node->key);
        free(del_node);
        ntt->items--;

        return 0;
    }

    return -5;
}

/* Point cursor to first item in tree */

struct ntt_node *c_ntt_first(struct ntt *ntt, struct ntt_c *c) {

    c->iter_index = 0;
    c->iter_next = (struct ntt_node *)NULL;
    return(c_ntt_next(ntt, c));
}

/* Point cursor to next iteration in tree */

struct ntt_node *c_ntt_next(struct ntt *ntt, struct ntt_c *c) {
    long index;
    struct ntt_node *node = c->iter_next;

    if (ntt == NULL) return NULL;

    if (node) {
        if (node != NULL) {
            c->iter_next = node->next;
            return (node);
        }
    }

    if (! node) {
        while (c->iter_index < ntt->size) {
            index = c->iter_index++;

            if (ntt->tbl[index]) {
                c->iter_next = ntt->tbl[index]->next;
                return(ntt->tbl[index]);
            }
        }
    }
    return((struct ntt_node *)NULL);
}

/* END NTT (Named Pointer Tree) Functions */

/* BEGIN Configuration Functions */

static const char *
get_hash_tbl_size(cmd_parms *cmd, void *dconfig, char *value) {
    long n = strtol(value, NULL, 0);

    if (n<=0) 
        hash_table_size = DEFAULT_HASH_TBL_SIZE;
    else 
        hash_table_size = n;

    return NULL;
}

static const char *
get_page_count(cmd_parms *cmd, void *dconfig, char *value) {
    long n = strtol(value, NULL, 0);
    if (n<=0) 
        page_count = DEFAULT_PAGE_COUNT;
    else
        page_count = n;

    return NULL;
}

static const char *
get_site_count(cmd_parms *cmd, void *dconfig, char *value) {
    long n = strtol(value, NULL, 0);
    if (n<=0) 
        site_count = DEFAULT_SITE_COUNT;
    else
        site_count = n;

    return NULL;
}

static const char *
get_page_interval(cmd_parms *cmd, void *dconfig, char *value) {
    long n = strtol(value, NULL, 0);
    if (n<=0) 
        page_interval = DEFAULT_PAGE_INTERVAL;
    else 
        page_interval = n;

    return NULL;
}

static const char *
get_site_interval(cmd_parms *cmd, void *dconfig, char *value) {
    long n = strtol(value, NULL, 0);
    if (n<=0) 
        site_interval = DEFAULT_SITE_INTERVAL;
    else
        site_interval = n;

  return NULL;
}

static const char *
get_blocking_period(cmd_parms *cmd, void *dconfig, char *value) {
    long n = strtol(value, NULL, 0);
    if (n<=0) 
        blocking_period = DEFAULT_BLOCKING_PERIOD;
    else 
        blocking_period = n;

    return NULL;
}

static const char *
get_log_dir(cmd_parms *cmd, void *dconfig, char *value) {
    if (value != NULL && value[0] != 0) {
        if (log_dir != NULL)
            free(log_dir);
        log_dir = strdup(value);
    }

    return NULL;
}

static const char *
get_email_notify(cmd_parms *cmd, void *dconfig, char *value) {
    if (value != NULL && value[0] != 0) {
        if (email_notify != NULL)
            free(email_notify);
        email_notify = strdup(value);
    }

    return NULL;
}

static const char *
get_sys_command(cmd_parms *cmd, void *dconfig, char *value) {
    if (value != NULL && value[0] != 0) {
        if (sys_command != NULL)
            free(sys_command);
        sys_command = strdup(value);
    }
 
    return NULL;
} 

static const char *whitelist(cmd_parms *cmd, void *mconfig, char *ip) {
    char entry[128];

    if (white_list == NULL) 
        white_list = ntt_create(53ul);
    snprintf(entry, sizeof(entry), "%s", ip);
    ntt_insert(white_list, entry, time(NULL));

    return NULL;
}

/* END Configuration Functions */

int is_whitelisted(const char *ip) {
    char hashkey[128];
    char octet[4][4];
    char *dip;
    char *oct;
    int i = 0;
                                                                                
    memset(octet, 0, 16);
    dip = strdup(ip);
    if (dip == NULL)
        return 0;
                                                                                
    oct = strtok(dip, ".");
    while(oct != NULL && i<4) {
        if (strlen(oct)<=3)
          strcpy(octet[i], oct);
        i++;
        oct = strtok(NULL, ".");
    }
    free(dip);
                                                                                
    /* Exact Match */
    snprintf(hashkey, sizeof(hashkey), "%s", ip);
    if (ntt_find(white_list, hashkey)!=NULL)
        return 1;
                                                                                
    /* IPv4 Wildcards */
    snprintf(hashkey, sizeof(hashkey), "%s.*.*.*", octet[0]);
    if (ntt_find(white_list, hashkey)!=NULL)
        return 1;
                                                                                
    snprintf(hashkey, sizeof(hashkey), "%s.%s.*.*", 
             octet[0], octet[1]);
    if (ntt_find(white_list, hashkey)!=NULL)
        return 1;

    snprintf(hashkey, sizeof(hashkey), "%s.%s.%s.*", 
             octet[0], octet[1], octet[2]);
    if (ntt_find(white_list, hashkey)!=NULL)
        return 1;

    /* No match */
    return 0;
}

static command_rec command_table[] = {

        { "DOSWhitelist", whitelist, NULL, RSRC_CONF, ITERATE,
        "Whitelist an IP or Wildcard. "},

	{ "DOSHashTableSize", get_hash_tbl_size, NULL, RSRC_CONF, TAKE1,
	"Set size of hash table. " },

	{ "DOSPageCount", get_page_count, NULL, RSRC_CONF, TAKE1,
	"Set maximum page hit count per interval. " },

	{ "DOSSiteCount", get_site_count, NULL, RSRC_CONF, TAKE1,
	"Set maximum site hit count per interval. " },

	{ "DOSPageInterval", get_page_interval, NULL, RSRC_CONF, TAKE1,
	"Set page interval. " },

	{ "DOSSiteInterval", get_site_interval, NULL, RSRC_CONF, TAKE1,
	"Set site interval. " }, 

	{ "DOSLogDir", get_log_dir, NULL, RSRC_CONF, TAKE1,
        "Set log dir. "},

	{ "DOSEmailNotify", get_email_notify, NULL, RSRC_CONF, TAKE1,
        "Set email notification. "},

	{ "DOSSystemCommand", get_sys_command, NULL, RSRC_CONF, TAKE1,
        "Set system command. "},

        { "DOSBlockingPeriod", get_blocking_period, NULL, RSRC_CONF, TAKE1,
        "Set blocking period for detected DoS IPs. "},

	{ NULL }
};

module MODULE_VAR_EXPORT evasive_module = {
    STANDARD_MODULE_STUFF,
    NULL,                              /* initializer */
    NULL,                              /* dir config creator */
    NULL,                              /* dir config merger */
    NULL,                              /* server config creator */
    NULL,                              /* server config merger */
    command_table,                     /* command table */
    NULL,                              /* handlers */
    NULL,                              /* filename translation */
    NULL,                              /* check_user_id */
    NULL,                              /* check auth */
    check_access,                      /* check access */
    NULL,                              /* type_checker */
    NULL,                              /* fixups */
    NULL,                              /* logger */
    NULL,                              /* header parser */
    evasive_child_init,                /* child_init */
    evasive_child_exit,                /* child_exit */
    NULL                               /* post read-request */
};

